base/tuple.jl

```diff
@@ -29,29 +29,43 @@ eltype{T,_}(::Type{NTuple{_,T}}) = T
 
 ## mapping ##
 
-ntuple(f::Function, n::Integer) =
+const ntuple = function(f, n::Integer)
     n<=0 ? () :
     n==1 ? (f(1),) :
     n==2 ? (f(1),f(2),) :
     n==3 ? (f(1),f(2),f(3),) :
-    n==4 ? (f(1),f(2),f(3),f(4),) :
-    n==5 ? (f(1),f(2),f(3),f(4),f(5),) :
-    tuple(ntuple(f,n-5)..., f(n-4), f(n-3), f(n-2), f(n-1), f(n))
-
-ntuple(f, ::Type{Val{0}}) = ()
-ntuple(f, ::Type{Val{1}}) = (f(1),)
-ntuple(f, ::Type{Val{2}}) = (f(1),f(2))
-ntuple(f, ::Type{Val{3}}) = (f(1),f(2),f(3))
-ntuple(f, ::Type{Val{4}}) = (f(1),f(2),f(3),f(4))
-ntuple(f, ::Type{Val{5}}) = (f(1),f(2),f(3),f(4),f(5))
-@generated function ntuple{N}(f, ::Type{Val{N}})
-    if !isa(N,Int)
-        :(throw(TypeError(:ntuple, "", Int, $(QuoteNode(N)))))
+    tuple([f(i) for i = 1:Int(n)]...)
```

Comment:
From the compiler standpoint it may be the easy part, but it's the one part that you can't do anything about from the Julia side short of writing a `@generated` function. (Inlining can be forced with `@inline`, `f` could be defined as a Functor.) I expect you noticed #13359. We're talking factors of 100 in terms of performance.